File: occasum-core/src/test/java/pw/stamina/occasum/properties/PropertyTests.java
package pw.stamina.occasum.properties;

import org.junit.jupiter.api.*;

import java.util.Random;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

public abstract class PropertyTests<T extends Property> {
    private static final Random RANDOM = new Random();
    protected T property;

    @BeforeEach
    void setupPropertyFromProvider() {
        property = provideProperty();
    }

    @Test
    @DisplayName("parseAndSet(null) -> throws PropertyParseException")
    void parseAndSet_givenNullInput_shouldThrowNullPointerException() {
        assertThrows(PropertyParseException.class, () -> property.parseAndSet(null));
    }

    @RepeatedTest(10)
    @DisplayName("reset() -> isDefault() returns true")
    void isDefault_afterResetHasBeenInvoked_shouldReturnTrue() {
        shufflePropertyValue(property, RANDOM);
        property.reset();
        assertTrue(property.isDefault());
    }

    protected abstract T provideProperty();

    protected abstract void shufflePropertyValue(T property, Random random);
}

File: occasum-core/src/test/kotlin/pw/stamina/occasum/properties/PropertyTests.kt
package pw.stamina.occasum.properties

import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.RepeatedTest
import java.util.*

abstract class PropertyTests<T : Property<*>> {
    protected lateinit var property: T

    @BeforeEach
    internal fun setupPropertyFromProvider() {
        property = provideProperty()
    }

    @RepeatedTest(10)
    @DisplayName("reset() -> isDefault() returns true")
    internal fun isDefault_afterResetHasBeenInvoked_shouldReturnTrue() {
        shufflePropertyValue(property, RANDOM)
        property.reset()
        assertTrue(property.isDefault)
    }

    protected abstract fun provideProperty(): T

    protected abstract fun shufflePropertyValue(property: T, random: Random)

    companion object {
        private val RANDOM = Random()
    }
}
